cpp/include/cuvs/neighbors/hnsw.hpp

```diff
@@ -0,0 +1,386 @@
+ *   hnsw::search(res, search_params, *index.get(), queries, neighbors, distances);
+ * @endcode
+ */
+void search(raft::resources const& res,
+            const search_params& params,
+            const index<int8_t>& idx,
+            raft::host_matrix_view<const int, int64_t, raft::row_major> queries,
+            raft::host_matrix_view<uint64_t, int64_t, raft::row_major> neighbors,
+            raft::host_matrix_view<float, int64_t, raft::row_major> distances);
+
+/**
+ * @brief De-serialize a CAGRA index saved to a file as an hnsw index
+ *
+ * @param[in] res raft resources
+ * @param[in] filename path to the file containing the serialized CAGRA index
+ * @param[in] dim dimensions of the training dataset
+ * @param[in] metric distance metric to search. Supported metrics ("L2Expanded", "InnerProduct")
+ * @param[out] index hnsw index
+ *
+ * Usage example:
+ * @code{.cpp}
+ *   // Build a CAGRA index
+ *   using namespace cuvs::neighbors;
+ *   // use default index parameters
+ *   cagra::index_params index_params;
+ *   // create and fill the index from a [N, D] dataset
+ *   auto index = cagra::build<float, uint32_t>(res, index_params, dataset);
+ *
+ *   // save a CAGRA index to a file
+ *   cagra::serialize(res, index, "index.bin");
+ *   // De-serialize a CAGRA index as a base-layer HNSW index using the filesystem
+ *   index<float>* hnsw_index = nullptr;
+ *   hnsw::deserialize(res, "index.bin", index->dim(), index->metric(), &hnsw_index);
+ *
+ *   // Delete index after use
+ *   delete hnsw_index;
+ * @endcode
+ */
+void deserialize(raft::resources const& res,
+                 const std::string& filename,
+                 int dim,
+                 cuvs::distance::DistanceType metric,
+                 index<float>** index);
+
+/**
+ * @brief De-serialize a CAGRA index saved to a file as an hnsw index
+ *
+ * @param[in] res raft resources
+ * @param[in] filename path to the file containing the serialized CAGRA index
+ * @param[in] dim dimensions of the training dataset
+ * @param[in] metric distance metric to search. Supported metrics ("L2Expanded", "InnerProduct")
+ * @param[out] index hnsw index
+ *
+ * Usage example:
+ * @code{.cpp}
+ *   // Build a CAGRA index
+ *   using namespace cuvs::neighbors;
+ *   // use default index parameters
+ *   cagra::index_params index_params;
+ *   // create and fill the index from a [N, D] dataset
+ *   auto index = cagra::build<uint8_t, uint32_t>(res, index_params, dataset);
+ *
+ *   // save a CAGRA index to a file
+ *   cagra::serialize(res, index, "index.bin");
+ *   // De-serialize a CAGRA index as a base-layer HNSW index using the filesystem
+ *   index<uint8_t>* hnsw_index = nullptr;
+ *   hnsw::deserialize(res, "index.bin", index->dim(), index->metric(), &hnsw_index);
+ *
+ *   // Delete index after use
+ *   delete hnsw_index;
+ * @endcode
+ */
+void deserialize(raft::resources const& res,
+                 const std::string& filename,
+                 int dim,
```

Comment:
Rather than have the caller specify the dimension and distance metric - can we read this from the serialized cagra index instead ? 

This means the user doesn't have to track extra metadata about a serialized index separately - and also will match the api of our other serialization methods

Reply:
There's 2 facets to this:
1. The serialized index is not a CAGRA index anymore, it's serialized in hnswlib format
2. Creating an hnswlib index and offloading deserialization to their APIs needs us to know the dimension beforehand

So, we could probably read this information from the index but it comes at the cost of abstraction. We would have to write our wrapper to read the index and information and then somehow plug that into hnswlib to avoid a second read.

Reply:
thanks for the explanation!

I took a look and it seems like while we could get the dim from the index (with some hacks), it isn't possible to get the distance metric from the hnswlib serialized index at all =(.

Reply:
Exactly, the serialized format is not intuitive in terms of the information that is written out!